_source/_posts/2023-09-15-java-spring-boot-microservices.md

```diff
@@ -0,0 +1,722 @@
+
+I had to remove the following dependencies from my original `api-gateway` project.
+
+```groovy
+implementation 'org.springframework.boot:spring-boot-starter-data-rest'
+implementation 'org.springframework.boot:spring-boot-starter-web'
+implementation 'org.springframework.cloud:spring-cloud-starter-openfeign'
+```
+
+And add Spring Cloud Gateway with Resilience4j dependencies:
+
+```groovy
+implementation 'org.springframework.cloud:spring-cloud-starter-circuitbreaker-reactor-resilience4j'
+implementation 'org.springframework.cloud:spring-cloud-starter-gateway'
+```
+
+Then, I moved `CoolCarController` to its own class and re-implemented it with WebClient.
+
+```java
+package com.example.apigateway.web;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.cloud.client.circuitbreaker.ReactiveCircuitBreaker;
+import org.springframework.cloud.client.circuitbreaker.ReactiveCircuitBreakerFactory;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.reactive.function.client.WebClient;
+import reactor.core.publisher.Flux;
+
+@RestController
+class CoolCarController {
+
+    Logger log = LoggerFactory.getLogger(CoolCarController.class);
+
+    private final WebClient.Builder webClientBuilder;
+    private final ReactiveCircuitBreaker circuitBreaker;
+
+    public CoolCarController(WebClient.Builder webClientBuilder,
+                             ReactiveCircuitBreakerFactory circuitBreakerFactory) {
+        this.webClientBuilder = webClientBuilder;
+        this.circuitBreaker = circuitBreakerFactory.create("circuit-breaker");
+    }
+
+    record Car(String name) {
+    }
+
+    @GetMapping("/cool-cars")
+    public Flux<Car> coolCars() {
+        return circuitBreaker.run(
+            webClientBuilder.build()
+                .get().uri("http://car-service/cars")
+                .retrieve().bodyToFlux(Car.class)
+                .filter(this::isCool),
+            throwable -> {
+                log.warn("Error making request to car service", throwable);
+                return Flux.empty();
+            });
+    }
+
+    private boolean isCool(Car car) {
+        return !car.name().equals("AMC Gremlin") &&
+            !car.name().equals("Triumph Stag") &&
+            !car.name().equals("Ford Pinto") &&
+            !car.name().equals("Yugo GV");
+    }
+}
+```
+
+In the `car-service` project, I had to switch from using Spring Data REST to handling it with a `@RestController` and `@GetMapping` annotation. I removed the `@RepositoryRestResource` annotation from `CarRepository` and added a `CarController` class.
+
```

Comment:
I can't figure out yet why you had to modify the car service.

Reply:
Because I was originally using `@RepositoryRestResource` on the repository, and it returns an `_embedded` element with the list of cars in there. I could not figure out how to get `WebClient` to parse it correctly. It worked fine with OpenFeight.

Reply:
I would add a note about that. Might be related to configuring [WebClient to speak hypermedia](https://docs.spring.io/spring-hateoas/docs/current/reference/html/#client.web-client) and 
["Spring Data Rest uses HAL to render responses"](https://docs.spring.io/spring-data/rest/docs/current/reference/html/#repository-resources.resource-discoverability)

Reply:
I didn't know changing the web client was this easy! Let me try it.

Reply:
I tried it and it doesn't work. As soon as I add `spring-boot-starter-hateoas` to the `build.gradle` in the API gateway project, it fails to start with:

```
Parameter 0 of constructor in com.example.apigateway.config.SecurityConfiguration required a bean of type 'org.springframework.security.oauth2.client.registration.ReactiveClientRegistrationRepository' that could not be found.

Action:

Consider defining a bean of type 'org.springframework.security.oauth2.client.registration.ReactiveClientRegistrationRepository' in your configuration.
```

If I remove the dependency, everything works. Here's the dependency coordinates:

```groovy
implementation 'org.springframework.boot:spring-boot-starter-hateoas'
```

Reply:
You can see my attempt in https://github.com/oktadev/auth0-java-microservices-examples/pull/7

Reply:
Add an issue to the Okta Spring Boot starter: https://github.com/okta/okta-spring-boot/issues/607

Reply:
Added note inline to address this issue.

Reply:
I just read the doc mentions [Spring MVC](https://docs.spring.io/spring-hateoas/docs/current/reference/html/#preface) but not reactive, could be related.